programs/validator-bonds/src/instructions/stake/merge_stake.rs

```diff
@@ -82,22 +87,18 @@ impl<'info> Merge<'info> {
 
         let destination_delegation = self.destination_stake.delegation();
```

Comment:
When the account had been previously delegated to a validator and then deactivated, is the `.delegation()` still `Some(_)`? Just double checking that the case when we have 2 de-activated accounts, we are indeed preventing their merging if they were previously a part of two different bonds

Reply:
That's a good point. I checked the code before and the test code works as expected. But until now I haven't been thinking that it could be changed, maybe in future.
I rechecked the solana code once again and in all code braches the deactivation means only to set the deactivation epoch but the delegation voter pubkey stays untouched. I added some verification check to test as well.
For current days I hope it could be considered safe (or let's see what the audit thinks).

But as you say currently the bonds merge operation depends strictly on the delegation field.
When thinking on some insurance out of the dependence of the Solana code I think that for Settlement funded stake account the securing element is the Settlement base `staker` authority. That seems to me fine.
For bonds funded stake account there is both authorities (staker and withdrawer) set to the same bond account, thus no way to differentiate. I'm going to change the code to require the `Stake` state (https://github.com/solana-labs/solana/blob/v1.18.3/sdk/program/src/stake/state.rs#L90) to not being tricked by some Solana enum addition in future. When the stake account is delegated in `Stake` state then there has to be at least some voter pubkey defined (it's not `Option` field`). I cannot be sure that they could not be somehow writing some specific pubkey (like `Pubkey.defualt`) but at least something for now.